rfcs/20200812-multidim-matmul/README.md

```diff
@@ -0,0 +1,188 @@
+  - Dependency on MKL binary.
+
+### Option 0.b Call GeMM in a loop
+Example: A = {3 x M x K} and B = {1 x K x N}
+```python
+for i in range(3) :
+  C[i] = GeMM(A[i], B)
+```
+
+- Pros:
+  - Uses existing GeMM impl.
+  - No maintenance/testing burden on oneDNN.
+- Cons: Performance may not be optimal.
+
+### Option 1 Extend MatMul primitive
+
+This primitive is similar to oneMKL [API](https://software.intel.com/content/www/us/en/develop/documentation/mkl-developer-reference-c/top/blas-and-sparse-blas-routines/blas-like-extensions/cblas-gemm-batch-strided.html)
+`cblas_?gemm_batch_strided`, as all matrices have fixed stride.
+
+**Broadcasting**
+  - Support implicit broadcasting, but keep the restriction
+  `ndims(A) == ndims(B) == ndims(C) == ndims(bias)` (Note: Unfortunately, this
+  means that the example at the beginning of this doc is not directly supported.
+  The tensor A must be manually expanded by user to match number of dims of B).
+  - Broadcasting on dims `M`, `N` and `K` is invalid and not supported.
+
+**Layouts**
+  - Support only plain and transposed layouts. For example, in a 5D tensor only
+  `abcde` and `abced` layouts are supported. Requires adding two tags (one for
+  plain and one for transpose) for each dim if not already present to the
+  library.
+
+**Maximum Dimensions**
+  - Currently oneDNN supports maximum of `12` dimensions for a tensor. This
+  implies that the maximum number of batch dimensions can be at most 10.
+
+**DNNL_RUNTIME_DIM_VAL and Broadcasting**
+  - The broadcasting consistency check is not done for the dimensions with
+  `DNNL_RUNTIME_DIM_VAL`. It is user responsibility to make sure the dims for
+  the tensors are valid (consistent with current runtime behavior of M, N, K
+  dims).
+  - Inconsistency with dims being "fully-defined" vs "runtime-defined" is
+  invalid (consistent with current behavior). Meaning, a "fully-defined" dim in
+  tensor A with a corresponding dim in B set as `DNNL_RUNTIME_DIM_VAL` is not
+  allowed. For example, `A` and `B` with dims set to `{3, 4, 5}` and
+  `{DNNL_RUNTIME_DIM_VAL, 4, 5}` respectively is invalid.
+
+**Zero points**: As is (scalar value).
+
+**Scales mask**: As is: (common scaling and per_oc (last dim, i.e.,
+`mask = 1 << (ndim -1)`)).
+
+**Primitive Cache**: No changes expected.
+
+**Benchdnn testing**
+We can keep the existing MatMul benchdnn driver for most part except `%desc%`.
+
+- Existing `%desc%`: An example shape for current `%desc%` looks as
+`mb2048m40n40k64`.
+
+- New `%desc%`:
+  - To support generic Tensor multiplication with support for arbitrary dims and
+  broadcast we can use `2048x16*x64:1x64x32[:2048x16x32]`, with optional `C`
+  dims. If `C` dims are not specified, they are calculated by driver. Here `A`,
+  `B` and `C` have shapes `{2048x16x64}`, `{1x64x32}` and `{2048x16x32}`
+  respectively.
+  - The asterisk `*` sign (in the example `16*`) represents runtime dim. The
+  `--runtime_X` options from the matmul driver will be dropped.
```

Comment:
The idea behind `--runtime_X` was to add an ability to test easily problems w/ or w/o runtime dims. I don't mind this change per se, but I think this would make the input files much more complex.

Summon @dzarukin .

Reply:
stupid idea to consider:
```
--bcast_dim_A=0 --bcast_dim_B=$mask mb128x3m30n20k10
```
here, `$mask` as usual: sum 2^d for d that we want to bcast, e.g.:
```
--bcast_dim_B=0 # no bcast
--bcast_dim_B=2 # bcast 3 (the second dim)
--bcast_dim_B=31 # bcast all 5 dims
```

Reply:
I think this is a good idea and I'll update the RFC accordingly.
(I was considering comma separated dims `2*,2x3,4` , but I think it will create way too many combinations than we desire.)

Reply:
Had a offline talk with @nivas-x86, and based on the case we want to cover there are two possible options:
1. Very general: `--runtime-A-mask=INT` and `--runtime-B-mask=INT`. This will let cover every broadcast case.
2. In case we don't want to support broadcast with run-time values, then more simple --runtime=M|N|K|<other_letters_for_different_dims> should be enough.

Reply:
> In case we don't want to support broadcast with run-time values

Could you please elaborate? Not sure I got the idea. Do you mean we will have O/P/Q/R/... letters for minibatch? Also, not sure I understand how broadcast is involved here?

Reply:
I like to keep it for general and with little more concise. So proposing `--runtime_dims_masks=<runtime_mask_a>:<runtime_mask_b>` . This can be used reduce the input files with `--runtime_dims_masks=<A_mask>:<B_mask>,<A_mask>:<B_mask>`.  (all masks are optional, 1 for a dim indicates runtime_dim.). Let me know your thoughts.

Reply:
> > In case we don't want to support broadcast with run-time values
> 
> Could you please elaborate? Not sure I got the idea. Do you mean we will have O/P/Q/R/... letters for minibatch? Also, not sure I understand how broadcast is involved here?

This I think regarding, the constraint in the RFC

`Inconsistency with dims being "fully-defined" vs "runtime-defined" is invalid (consistent with current behavior). Meaning, a "fully-defined" dim in tensor A with a corresponding dim in B set as DNNL_RUNTIME_DIM_VAL is not allowed. For example, A and B with dims set to {3, 4, 5} and {DNNL_RUNTIME_DIM_VAL, 4, 5} respectively is invalid`

He had a thought of supporting this (now or in future), and runtime_dim_value can be either `1` or `3` can be chosen at runtime to either broadcast or not. (@dzarukin  hope I'm right about your intent)

Reply:
So, there are couple things on the plate we need to agree on about benchdnn support.

a) Broadcasting. There are two options:
- `mbY[xY[...]]mMnNkK` - putting multi-batch under `mb` entry, keeping `desc_t` intact. It means to specify broadcast one needs to create a flag `--mb-broadcast=mask`.
  * Pros: No need to break existing problem description concept.
  * Cons: if b2), it happens that friend two (or three) flags together and keep the consistency between them may be problematic.
  * Observation: A hybrid model of dims_t and desc, may cause some confusion; seems to work good enough with b1).
- `Y[xY[...]]xMxK:Z[xZ[...]]xKxN` - moving to dims_t with a little ambiguity in K (which will have to coincide for both inputs). This way mb for B is completely independent and broadcast semantics can be expressed the same way as for the binary driver.
  * Pros: Keeping broadcast semantics aligned between [two] drivers [that support a broadcast].
  * Cons: New concept for the driver, need to keep backward compatibility for some time + updating input files.
  * Observation: Seems to work nice with any b1) or b2).

b) Runtime dimensions. There are also two proposals:
- `mM*nN*kK* / M*xK*:K*xN*`- setting a star specifier (meaning run-time value) in a descriptor/dims, no additional flags support.
  * Pros: no additional flags burden;
  * Cons: lack of flexibility, to run a problem with and without runtime will require two problems instead of a single value in an option; additional parser modification is required to process a run-time value and keep parsed info until creating a memory descriptor.
- `--runtime-A=mask` and `--runtime-B=mask` / `--runtime=A_mask:B_mask` - adding flags for each matrix to specify what values should be run-time.
  * Pros are Cons of b1) and vice versa.

To my mind, only two options are viable: 1a)-2a) as a quick way to implement testing in case we don't expect this feature will get any further development, thus, no benchdnn touching in the future in matmul. Or 1b)-2b) as a general extension which will not require any changes in future if such multi-batch direction will get further improvements. I prefer 1b)-2b) over 1a)-2a), but I don't have enough information to say 1a)-2a) should not be implemented. 1a)-2b) will complicate a driver API much. 1b)-2a) doesn't make much sense due to new desc concept itself doesn't bring much value. So it's either a lazy approach or general thinking-about-the-future approach.

Reply:
Thanks @dzarukin . I agree with you (1b-2b) and fortunately my current implementation already follows this approach.